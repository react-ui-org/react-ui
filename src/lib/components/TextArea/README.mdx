---
name: TextArea
menu: 'Inputs'
route: /components/text-area
---

# TextArea

TextArea enables users to add longer text to a form.

import {
  Playground,
  Props,
} from 'docz'
import { TextArea } from './TextArea'

## Basic Usage

To implement the TextArea component, you need to import it first:

```js
import { TextArea } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <TextArea label="Your feedback" />
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Use **short and descriptive labels**, ideally nouns rather than seemingly
  polite phrases like _Please enter your message_. Short labels will help your
  users accomplish their task faster.

- Only make the TextArea's label invisible when there is **another visual
  clue** to guide users through filling the input. Using the first option as
  label is not recommended either ‚Äî it disappears once user makes their choice.

- When a short label is not enough, use **help texts to guide users** before
  they enter anything.

- Optionally you can **pre-fill your text fields with a placeholder** to give
  users an example of what they're expected to fill in. Just remember the
  placeholder disappears once a value is entered so don't put anything important
  there. Usually, en example value starting with ‚ÄúEg.‚Äù works best.

- Use **clear, calm error messages** when there's a problem with what they
  entered. Be positive and focus on solutions to make the error message helpful.

- Ensure the height of a text area **fits within mobile screen sizes.**

## Design Variants

To satisfy the design requirements of your project, all input fields in React UI
come in **two design variants** to choose from: outline and filled. Both can be
further [customized](#theming) with CSS custom properties.

<Playground>
  <TextArea
    label="Your feedback"
  />
  <TextArea
    label="Your feedback"
    variant="filled"
  />
</Playground>

## Sizes

Aside from the default (medium) size, two additional sizes are available: small
and large.

<Playground>
  <TextArea
    label="Your feedback"
    size="small"
  />
  <TextArea
    label="Your feedback"
  />
  <TextArea
    label="Your feedback"
    size="large"
  />
  <TextArea
    label="Your feedback"
    size="small"
    variant="filled"
  />
  <TextArea
    label="Your feedback"
    variant="filled"
  />
  <TextArea
    label="Your feedback"
    size="large"
    variant="filled"
  />
</Playground>

Full-width fields span the full width of a parent:

<Playground>
  <TextArea
    label="Your feedback"
    fullWidth
  />
  <TextArea
    label="Your feedback"
    variant="filled"
    fullWidth
  />
</Playground>

## Input Size

The default width of all inputs is 240 px, and it can be
[customized](/customize/theming/overview) with a CSS custom property. However, you
can also **control the size of individual text areas** using the `rows` and
`cols` properties. Additionally, text areas are vertically resizable by users.

üëâ Remember that the `cols` and `rows` HTML attributes **do not limit on how
many characters** the user can enter. Use the
[`maxlength` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea#Min_and_max_length)
to achieve that effect.

<Playground>
  <TextArea
    label="Your feedback"
    cols={50}
    rows={6}
  />
  <TextArea
    label="Your feedback"
    variant="filled"
    cols={50}
    rows={6}
  />
</Playground>

## Forwarding HTML Attributes

When you want to improve the accessibility of your text fields even further, you
can **add whatever HTML attribute you like.** All attributes that don't
interfere with [component's API](#api) are forwarded to the native HTML input.

<Playground>
  <TextArea
    label="Address"
    autoComplete="street-address"
    minLength={3}
    maxLength={80}
  />
  <TextArea
    label="Address"
    variant="filled"
    autoComplete="street-address"
    minLength={3}
    maxLength={80}
  />
</Playground>

üëâ Refer to the MDN reference for the full list of
[supported attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea#Attributes).

## Invisible Label

In some cases, it may be convenient to visually hide the field label. The label
remains accessible to assistive technologies.

While it may be acceptable for login screens with just a few fields or other
simple forms, it's dangerous to hide labels from users in most cases. Keep in
mind you should **provide another visual clue** so users know what to fill into
the input.

<Playground>
  <TextArea
    label="Feedback"
    isLabelVisible={false}
    placeholder="Eg. What I liked the best was‚Ä¶"
  />
  <TextArea
    label="Feedback"
    isLabelVisible={false}
    variant="filled"
    placeholder="Eg. What I liked the best was‚Ä¶"
  />
</Playground>

## Horizontal Layout

The default vertical layout is very easy to use and work with. However, there
are situations where horizontal layout suits better ‚Äî and that's why React UI
supports this kind of layout as well.

<Playground>
  <TextArea
    label="Your feedback"
    layout="horizontal"
  />
  <TextArea
    label="Your feedback"
    layout="horizontal"
    variant="filled"
  />
  <TextArea
    label="Your feedback"
    layout="horizontal"
    fullWidth
  />
  <TextArea
    label="Your feedback"
    layout="horizontal"
    variant="filled"
    fullWidth
  />
  <TextArea
    isLabelVisible={false}
    label="Your feedback"
    layout="horizontal"
    placeholder="Your feedback"
    fullWidth
  />
  <TextArea
    isLabelVisible={false}
    label="Your feedback"
    layout="horizontal"
    placeholder="Your feedback"
    variant="filled"
    fullWidth
  />
</Playground>

## Help Text

You may provide an additional help text to clarify how the input should be
filled.

<Playground>
  <TextArea
    label="Your feedback"
    helpText="Why would you recommend us?"
  />
  <TextArea
    label="Your feedback"
    variant="filled"
    helpText="Why would you recommend us?"
  />
  <TextArea
    label="Your feedback"
    layout="horizontal"
    helpText="Why would you recommend us?"
  />
  <TextArea
    label="Your feedback"
    layout="horizontal"
    helpText="Why would you recommend us?"
    variant="filled"
  />
  <TextArea
    label="Your feedback"
    layout="horizontal"
    helpText="Why would you recommend us?"
    fullWidth
  />
  <TextArea
    label="Your feedback"
    layout="horizontal"
    variant="filled"
    helpText="Why would you recommend us?"
    fullWidth
  />
</Playground>

## States

### Validation States

Validation states visually present the result of validation of the input. You
should always **provide a help text for states other than valid** so users know
what happened and what action they should take or what options they have.

<Playground>
  <TextArea
    label="Your feedback"
    validationState="valid"
    validationText="Looks good!"
    value="Like that!"
    required
  />
  <TextArea
    label="Your feedback"
    validationState="warning"
    validationText="The feedback should be at least 3 characters long."
    value=".."
    required
  />
  <TextArea
    label="Your feedback"
    validationState="invalid"
    validationText="Please share your feedback with us."
    required
  />
  <TextArea
    label="Your feedback"
    validationState="valid"
    validationText="Looks good!"
    variant="filled"
    value="Like that!"
    required
  />
  <TextArea
    label="Your feedback"
    validationState="warning"
    validationText="The feedback should be at least 3 characters long."
    variant="filled"
    value=".."
    required
  />
  <TextArea
    label="Your feedback"
    validationState="invalid"
    validationText="Please share your feedback with us."
    variant="filled"
    required
  />
</Playground>

### Disabled State

It's possible to disable the whole input.

<Playground>
  <TextArea
    label="Your feedback"
    disabled
  />
  <TextArea
    label="Your feedback"
    variant="filled"
    disabled
  />
</Playground>

## API

In addition to the options below, you can specify [React synthetic events] or
any custom HTML attributes that do not interfere with the API, and they will be
passed to the `textarea` HTML element. This enables making the component
interactive and helps improve its [accessibility](#forwarding-html-attributes).

<Props table of={TextArea} />

## Theming

Head to [Forms Theming](/customize/theming/forms) to see shared form theming
options.

[React synthetic events]: https://reactjs.org/docs/events.html
