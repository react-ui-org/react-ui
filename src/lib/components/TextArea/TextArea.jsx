import PropTypes from 'prop-types';
import React, { useContext } from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootSizeClassName } from '../_helpers/getRootSizeClassName';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { resolveContextOrProp } from '../_helpers/resolveContextOrProp';
import { transferProps } from '../_helpers/transferProps';
import { FormLayoutContext } from '../FormLayout';
import withForwardedRef from '../withForwardedRef';
import styles from './TextArea.scss';

export const TextArea = ({
  cols,
  disabled,
  forwardedRef,
  fullWidth,
  helpText,
  id,
  isLabelVisible,
  label,
  layout,
  placeholder,
  required,
  rows,
  size,
  validationState,
  validationText,
  value,
  variant,
  ...restProps
}) => {
  const context = useContext(FormLayoutContext);

  return (
    <label
      className={classNames(
        styles.root,
        fullWidth && styles.isRootFullWidth,
        context && styles.isRootInFormLayout,
        resolveContextOrProp(context && context.layout, layout) === 'horizontal'
          ? styles.rootLayoutHorizontal
          : styles.rootLayoutVertical,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootSizeClassName(size, styles),
        getRootValidationStateClassName(validationState, styles),
        variant === 'filled' ? styles.rootVariantFilled : styles.rootVariantOutline,
      )}
      htmlFor={id}
      id={id && `${id}__label`}
    >
      <div
        className={classNames(
          styles.label,
          !isLabelVisible && styles.isLabelHidden,
        )}
        id={id && `${id}__labelText`}
      >
        {label}
      </div>
      <div className={styles.field}>
        <div className={styles.inputContainer}>
          <textarea
            {...transferProps(restProps)}
            className={styles.input}
            cols={cols}
            disabled={disabled}
            id={id}
            placeholder={placeholder}
            ref={forwardedRef}
            required={required}
            rows={rows}
            value={value}
          />
          {variant === 'filled' && (
            <div className={styles.bottomLine} />
          )}
        </div>
        {helpText && (
          <div
            className={styles.helpText}
            id={id && `${id}__helpText`}
          >
            {helpText}
          </div>
        )}
        {validationText && (
          <div
            className={styles.validationText}
            id={id && `${id}__validationText`}
          >
            {validationText}
          </div>
        )}
      </div>
    </label>
  );
};

TextArea.defaultProps = {
  cols: null,
  disabled: false,
  forwardedRef: undefined,
  fullWidth: false,
  helpText: null,
  id: undefined,
  isLabelVisible: true,
  layout: 'vertical',
  placeholder: null,
  required: false,
  rows: 3,
  size: 'medium',
  validationState: null,
  validationText: null,
  value: undefined,
  variant: 'outline',
};

TextArea.propTypes = {
  /**
   * The number of visible text columns for the control.
   */
  cols: PropTypes.number,
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Reference forwarded to the `textarea` element.
   */
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    // eslint-disable-next-line react/forbid-prop-types
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  /**
   * If `true`, the field will span the full width of its parent.
   */
  fullWidth: PropTypes.bool,
  /**
   * Optional help text.
   */
  helpText: PropTypes.node,
  /** ID of the input HTML element. It also serves as a prefix for nested elements:
   * * `<ID>__label`
   * * `<ID>__labelText`
   * * `<ID>__helpText`
   * * `<ID>__validationText`
   */
  id: PropTypes.string,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * Text field label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Layout of the field.
   *
   * Ignored if the component is rendered within `FormLayout` component
   * as the value is inherited in such case.
   */
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  /**
   * Optional example value.
   */
  placeholder: PropTypes.string,
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * The number of visible text lines for the control.
   */
  rows: PropTypes.number,
  /**
   * Size of the field.
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Validation message to be displayed.
   */
  validationText: PropTypes.node,
  /**
   * Value of the input.
   */
  value: PropTypes.string,
  /**
   * Design variant of the field, further customizable with CSS custom properties.
   */
  variant: PropTypes.oneOf(['filled', 'outline']),
};

export const TextAreaWithContext = withForwardedRef(withProviderContext(TextArea, 'TextArea'));

export default TextAreaWithContext;
