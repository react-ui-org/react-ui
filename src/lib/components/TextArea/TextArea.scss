@use "../../styles/tools/form-fields/box-field-elements";
@use "../../styles/tools/form-fields/box-field-layout";
@use "../../styles/tools/form-fields/box-field-sizes";
@use "../../styles/tools/form-fields/foundation";
@use "../../styles/tools/form-fields/variants";
@use "../../styles/tools/accessibility";

// Foundation
.root {
    @include foundation.root();
}

.label {
    @include foundation.label();
}

.inputContainer {
    @include box-field-elements.input-container();
}

.input {
    @include box-field-elements.input();
    @include box-field-elements.input-textarea();
}

.bottomLine {
    @include box-field-elements.bottom-line();
}

.helpText,
.validationText {
    @include foundation.help-text();
}

.isRootRequired .label {
    @include foundation.label-required();
}

// Visual variants
.rootVariantFilled {
    @include variants.visual(box, $variant: filled);
}

.rootVariantOutline {
    @include variants.visual(box, $variant: outline);
}

// States
.isRootStateInvalid {
    @include variants.validation(invalid);
}

.isRootStateValid {
    @include variants.validation(valid);
}

.isRootStateWarning {
    @include variants.validation(warning);
}

// Invisible label
.isLabelHidden {
    @include accessibility.hide-text();
}

// Layouts
.rootLayoutVertical,
.rootLayoutHorizontal {
    @include box-field-layout.vertical();
}

.rootLayoutHorizontal {
    @include box-field-layout.horizontal();
}

.isRootFullWidth {
    @include box-field-layout.full-width();
}

.isRootInFormLayout {
    @include box-field-layout.in-form-layout();
}

// Sizes
.rootSizeSmall {
    @include box-field-sizes.size(small, $is-multiline: true);
}

.rootSizeMedium {
    @include box-field-sizes.size(medium, $is-multiline: true);
}

.rootSizeLarge {
    @include box-field-sizes.size(large, $is-multiline: true);
}
