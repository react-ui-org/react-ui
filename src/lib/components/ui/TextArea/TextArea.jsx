import PropTypes from 'prop-types';
import React from 'react';
import transferProps from '../../../utils/transferProps';
import withForwardedRef from '../withForwardedRef';
import styles from './TextArea.scss';

export const TextArea = (props) => {
  let labelVisibilityClass = '';
  let rootFullWidthClass = '';
  let rootInFormLayoutClass = '';
  let rootLayoutClass = '';
  let rootRequiredClass = '';
  let rootSizeClass = '';
  let rootValidationStateClass = '';
  let rootVariantClass = '';

  if (!props.isLabelVisible) {
    labelVisibilityClass = styles.isLabelHidden;
  }

  if (props.fullWidth) {
    rootFullWidthClass = styles.isRootFullWidth;
  }

  if (props.inFormLayout) {
    rootInFormLayoutClass = styles.isRootInFormLayout;
  }

  if (props.layout === 'horizontal') {
    rootLayoutClass = styles.rootLayoutHorizontal;
  } else if (props.layout === 'vertical') {
    rootLayoutClass = styles.rootLayoutVertical;
  }

  if (props.required) {
    rootRequiredClass = styles.isRootRequired;
  }

  if (props.size === 'small') {
    rootSizeClass = styles.rootSizeSmall;
  } else if (props.size === 'medium') {
    rootSizeClass = styles.rootSizeMedium;
  } else if (props.size === 'large') {
    rootSizeClass = styles.rootSizeLarge;
  }

  if (props.validationState === 'invalid') {
    rootValidationStateClass = styles.isRootStateInvalid;
  } else if (props.validationState === 'valid') {
    rootValidationStateClass = styles.isRootStateValid;
  } else if (props.validationState === 'warning') {
    rootValidationStateClass = styles.isRootStateWarning;
  }

  if (props.variant === 'filled') {
    rootVariantClass = styles.rootVariantFilled;
  } else if (props.variant === 'outline') {
    rootVariantClass = styles.rootVariantOutline;
  }

  const propsToTransfer = transferProps(
    props,
    ['changeHandler', 'cols', 'disabled', 'fullWidth', 'helperText', 'id', 'inFormLayout', 'isLabelVisible',
      'label', 'layout', 'placeholder', 'required', 'rows', 'size', 'validationState', 'value', 'variant'],
  );

  return (
    <label
      className={(`
        ${styles.root}
        ${rootFullWidthClass}
        ${rootInFormLayoutClass}
        ${rootLayoutClass}
        ${rootRequiredClass}
        ${rootSizeClass}
        ${rootValidationStateClass}
        ${rootVariantClass}
      `).trim()}
      htmlFor={props.id}
      id={`${props.id}__label`}
    >
      <div
        className={(`
          ${styles.label}
          ${labelVisibilityClass}
        `).trim()}
        id={`${props.id}__labelText`}
      >
        {props.label}
      </div>
      <div className={styles.field}>
        <div className={styles.inputContainer}>
          <textarea
            {...propsToTransfer}
            className={styles.input}
            cols={props.cols}
            disabled={props.disabled}
            id={props.id}
            onChange={props.changeHandler}
            placeholder={props.placeholder}
            ref={props.forwardedRef}
            required={props.required}
            rows={props.rows}
            value={props.value}
          />
          {props.variant === 'filled' && (
            <div className={styles.bottomLine} />
          )}
        </div>
        {props.helperText && (
          <div
            className={styles.helperText}
            id={`${props.id}__helperText`}
          >
            {props.helperText}
          </div>
        )}
      </div>
    </label>
  );
};

TextArea.defaultProps = {
  changeHandler: null,
  cols: null,
  disabled: false,
  forwardedRef: undefined,
  fullWidth: false,
  helperText: null,
  inFormLayout: false,
  isLabelVisible: true,
  layout: 'vertical',
  placeholder: null,
  required: false,
  rows: 3,
  size: 'medium',
  validationState: null,
  value: undefined,
  variant: 'outline',
};

TextArea.propTypes = {
  changeHandler: PropTypes.func,
  cols: PropTypes.number,
  disabled: PropTypes.bool,
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  fullWidth: PropTypes.bool,
  helperText: PropTypes.string,
  id: PropTypes.string.isRequired,
  inFormLayout: PropTypes.bool,
  isLabelVisible: PropTypes.bool,
  label: PropTypes.string.isRequired,
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  rows: PropTypes.number,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  value: PropTypes.string,
  variant: PropTypes.oneOf(['filled', 'outline']),
};

export default withForwardedRef(TextArea);
