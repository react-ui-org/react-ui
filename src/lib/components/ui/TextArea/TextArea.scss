@import '../../../styles/tools/forms/foundation';
@import '../../../styles/tools/forms/layouts';
@import '../../../styles/tools/forms/relationships';
@import '../../../styles/tools/forms/sizes';
@import '../../../styles/tools/forms/states';
@import '../../../styles/tools/forms/variants';
@import '../../../styles/tools/accessibility';

// Foundation
.root {
  @include form-field();
  @include form-field-horizontal-neighbor();
}

.label {
  @include form-field-label();
}

.inputContainer {
  @include form-field-input-container();
}

.input {
  @include form-field-input();
  @include form-field-input-textarea();
}

.bottomLine {
  @include form-field-bottom-line();
}

.helpText,
.validationText {
  @include form-field-help-text();
}

// Variants
.rootVariantFilled .input {
  @include form-field-input-variant(filled);
}

.rootVariantOutline .input {
  @include form-field-input-variant(outline);
}

// States
.isRootRequired .label {
  @include form-field-label-required();
}

.isRootStateInvalid {
  @include form-field-state(invalid);
}

.isRootStateValid {
  @include form-field-state(valid);
}

.isRootStateWarning {
  @include form-field-state(warning);
}

// Invisible label
.isLabelHidden {
  @include accessibility-hide-text();
}

// Layouts
.rootLayoutHorizontal {
  @include form-field-layout-horizontal();
}

.rootLayoutVertical {
  @include form-field-layout-vertical();
}

.isRootFullWidth {
  @include form-field-full-width();
  @include form-field-vertical-neighbor();
}

.isRootInFormLayout {
  @include form-field-in-form-layout();
}

// Sizes
.rootSizeSmall {
  @include form-field-size(small, $is-multiline: true);
}

.rootSizeMedium {
  @include form-field-size(medium, $is-multiline: true);
}

.rootSizeLarge {
  @include form-field-size(large, $is-multiline: true);
}
