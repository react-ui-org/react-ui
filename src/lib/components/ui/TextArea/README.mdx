---
name: TextArea
menu: 'Components: UI'
route: /components/ui/text-area
---

# TextArea

TextArea enables users to add longer text to a form.

import {
  Playground,
  Props,
} from 'docz'
import { TextArea } from './TextArea'

## Basic Usage

To implement the TextArea component, you need to import it first:

```js
import { TextArea } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <TextArea id="my-text-area" label="Your feedback" />
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Use **short and descriptive labels**, ideally nouns rather than seemingly
  polite phrases like _Please enter your message_. Short labels will help your
  users accomplish their task faster.

- Only make the TextArea's label invisible when there is **another visual
  clue** to guide users through filling the input. Using the first option as
  label is not recommended either ‚Äî it disappears once user makes their choice.

- When a short label is not enough, use **help texts to guide users** before
  they enter anything.

- Optionally you can **pre-fill your text fields with a placeholder** to give
  users an example of what they're expected to fill in. Just remember the
  placeholder disappears once a value is entered so don't put anything important
  there. Usually, en example value starting with ‚ÄúEg.‚Äù works best.

- Use **clear, calm error messages** when there's a problem with what they
  entered. Be positive and focus on solutions to make the error message helpful.

- Ensure the height of a text area **fits within mobile screen sizes.**

## Design Variants

To satisfy the design requirements of your project, all input fields in React UI
come in **two design variants** to choose from: outline and filled. Both can be
further [customized](/customize/theming) with CSS custom properties.

<Playground>
  <TextArea
    id="my-text-area-outline"
    label="Your feedback"
  />
  <TextArea
    id="my-text-area-filled"
    label="Your feedback"
    variant="filled"
  />
</Playground>

## Sizes

Aside from the default (medium) size, two additional sizes are available: small
and large.

<Playground>
  <TextArea
    id="my-text-area-outline-small"
    label="Your feedback"
    size="small"
  />
  <TextArea
    id="my-text-area-outline-medium"
    label="Your feedback"
  />
  <TextArea
    id="my-text-area-outline-large"
    label="Your feedback"
    size="large"
  />
  <TextArea
    id="my-text-area-filled-small"
    label="Your feedback"
    size="small"
    variant="filled"
  />
  <TextArea
    id="my-text-area-filled-medium"
    label="Your feedback"
    variant="filled"
  />
  <TextArea
    id="my-text-area-filled-large"
    label="Your feedback"
    size="large"
    variant="filled"
  />
</Playground>

Full-width fields span the full width of a parent:

<Playground>
  <TextArea
    id="my-text-area-outline-full-width"
    label="Your feedback"
    fullWidth
  />
  <TextArea
    id="my-text-area-filled-full-width"
    label="Your feedback"
    variant="filled"
    fullWidth
  />
</Playground>

## Input Size

The default width of all inputs is 240 px, and it can be
[customized](/customize/theming) with a CSS custom property. However, you
can also **control the size of individual text areas** using the `rows` and
`cols` properties. Additionally, text areas are vertically resizable by users.

üëâ Remember that the `cols` and `rows` HTML attributes **do not limit on how
many characters** the user can enter. Use the
[`maxlength` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea#Min_and_max_length)
to achieve that effect.

<Playground>
  <TextArea
    id="my-text-area-outline-input-size"
    label="Your feedback"
    cols={50}
    rows={6}
  />
  <TextArea
    id="my-text-area-filled-input-size"
    label="Your feedback"
    variant="filled"
    cols={50}
    rows={6}
  />
</Playground>

## Forwarding HTML Attributes

When you want to improve the accessibility of your text fields even further, you
can **add whatever HTML attribute you like.** All attributes that don't
interfere with [component's API](#api) are forwarded to the native HTML input.

<Playground>
  <TextArea
    id="my-text-area-outline-forwarding"
    label="Address"
    autoComplete="street-address"
    minLength={3}
    maxLength={80}
  />
  <TextArea
    id="my-text-area-filled-forwarding"
    label="Address"
    variant="filled"
    autoComplete="street-address"
    minLength={3}
    maxLength={80}
  />
</Playground>

üëâ Refer to the MDN reference for the full list of
[supported attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea#Attributes).

## Invisible Label

In some cases, it may be convenient to visually hide the field label. The label
remains accessible to assistive technologies.

While it may be acceptable for login screens with just a few fields or other
simple forms, it's dangerous to hide labels from users in most cases. Keep in
mind you should **provide another visual clue** so users know what to fill into
the input.

<Playground>
  <TextArea
    id="my-text-area-outline-hidden-label"
    label="Feedback"
    isLabelVisible={false}
    placeholder="Eg. What I liked the best was‚Ä¶"
  />
  <TextArea
    id="my-text-area-filled-hidden-label"
    label="Feedback"
    isLabelVisible={false}
    variant="filled"
    placeholder="Eg. What I liked the best was‚Ä¶"
  />
</Playground>

## Horizontal Layout

The default vertical layout is very easy to use and work with. However, there
are situations where horizontal layout suits better ‚Äî and that's why React UI
supports this kind of layout as well.

<Playground>
  <TextArea
    id="my-text-area-outline-horizontal"
    label="Your feedback"
    layout="horizontal"
  />
  <TextArea
    id="my-text-area-filled-horizontal"
    label="Your feedback"
    layout="horizontal"
    variant="filled"
  />
  <TextArea
    id="my-text-area-outline-horizontal-full-width"
    label="Your feedback"
    layout="horizontal"
    fullWidth
  />
  <TextArea
    id="my-text-area-filled-horizontal-full-width"
    label="Your feedback"
    layout="horizontal"
    variant="filled"
    fullWidth
  />
  <TextArea
    id="my-text-area-outline-horizontal-full-width-invisible-label"
    isLabelVisible={false}
    label="Your feedback"
    layout="horizontal"
    placeholder="Your feedback"
    fullWidth
  />
  <TextArea
    id="my-text-area-filled-horizontal-full-width-invisible-label"
    isLabelVisible={false}
    label="Your feedback"
    layout="horizontal"
    placeholder="Your feedback"
    variant="filled"
    fullWidth
  />
</Playground>

## Help Text

You may provide an additional help text to clarify how the input should be
filled.

<Playground>
  <TextArea
    id="my-text-area-outline-help-text"
    label="Your feedback"
    helpText="Why would you recommend us?"
  />
  <TextArea
    id="my-text-area-filled-help-text"
    label="Your feedback"
    variant="filled"
    helpText="Why would you recommend us?"
  />
  <TextArea
    id="my-text-area-outline-help-text-horizontal"
    label="Your feedback"
    layout="horizontal"
    helpText="Why would you recommend us?"
  />
  <TextArea
    id="my-text-area-filled-help-text-horizontal"
    label="Your feedback"
    layout="horizontal"
    helpText="Why would you recommend us?"
    variant="filled"
  />
  <TextArea
    id="my-text-area-outline-help-text-horizontal-full-width"
    label="Your feedback"
    layout="horizontal"
    helpText="Why would you recommend us?"
    fullWidth
  />
  <TextArea
    id="my-text-area-filled-help-text-horizontal-full-width"
    label="Your feedback"
    layout="horizontal"
    variant="filled"
    helpText="Why would you recommend us?"
    fullWidth
  />
</Playground>

## States

### Validation States

Validation states visually present the result of validation of the input. You
should always **provide a help text for states other than valid** so users know
what happened and what action they should take or what options they have.

<Playground>
  <TextArea
    id="my-text-area-outline-success"
    label="Your feedback"
    validationState="valid"
    validationText="Looks good!"
    changeHandler={() => {}}
    value="Like that!"
    required
  />
  <TextArea
    id="my-text-area-outline-warning"
    label="Your feedback"
    validationState="warning"
    validationText="The feedback should be at least 3 characters long."
    value=".."
    changeHandler={() => {}}
    required
  />
  <TextArea
    id="my-text-area-outline-error"
    label="Your feedback"
    validationState="invalid"
    validationText="Please share your feedback with us."
    required
  />
  <TextArea
    id="my-text-area-filled-success"
    label="Your feedback"
    validationState="valid"
    validationText="Looks good!"
    variant="filled"
    changeHandler={() => {}}
    value="Like that!"
    required
  />
  <TextArea
    id="my-text-area-filled-warning"
    label="Your feedback"
    validationState="warning"
    validationText="The feedback should be at least 3 characters long."
    variant="filled"
    changeHandler={() => {}}
    value=".."
    required
  />
  <TextArea
    id="my-text-area-filled-error"
    label="Your feedback"
    validationState="invalid"
    validationText="Please share your feedback with us."
    variant="filled"
    required
  />
</Playground>

### Disabled State

It's possible to disable the whole input.

<Playground>
  <TextArea
    id="my-text-area-outline-disabled"
    label="Your feedback"
    disabled
  />
  <TextArea
    id="my-text-area-filled-disabled"
    label="Your feedback"
    variant="filled"
    disabled
  />
</Playground>

## API

In addition to the options below, you can add any custom attributes that do not
interfere with the API, and they will be passed to the `textarea` HTML element.
This is useful mainly to improve component's
[accessibility](#forwarding-html-attributes).

<Props of={TextArea} />

## Theming

Head to [Forms Theming](/customize/theming#forms).

---

Next: [Toggle ‚Üí](/components/ui/toggle)
