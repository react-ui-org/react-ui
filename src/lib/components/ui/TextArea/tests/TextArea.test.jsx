import React from 'react';
import {
  shallow,
  mount,
} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import { TextArea } from '../TextArea';

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<TextArea label="label" id="test" />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with hidden label', () => {
    const tree = shallow(<TextArea
      id="test"
      label="label"
      isLabelVisible={false}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<TextArea
      cols={10}
      id="test"
      helperText="some help"
      isLabelVisible={false}
      label="label"
      layout="horizontal"
      placeholder="placeholder"
      rows={5}
      validationState="invalid"
      value="value"
      variant="filled"
      size="large"
      fullWidth
      disabled
      required
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls changeHandler()', () => {
    const spy = sinon.spy();
    const component = mount(<TextArea
      id="test"
      label="label"
      changeHandler={spy}
    />);

    component
      .find('textarea')
      .simulate('change', {
        preventDefault: () => {},
        target: { value: 'test' },
      });
    expect(spy.calledOnce).toEqual(true);
  });
});
