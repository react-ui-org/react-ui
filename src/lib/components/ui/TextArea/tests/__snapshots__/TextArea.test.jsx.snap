// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<label
  className="root
        
        
        rootLayoutVertical
        
        rootSizeMedium
        
        rootVariantOutline"
  htmlFor="test"
  id="test__label"
>
  <div
    className="label"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="field"
  >
    <div
      className="inputContainer"
    >
      <textarea
        className="input"
        cols={null}
        disabled={false}
        id="test"
        onChange={null}
        placeholder={null}
        required={false}
        rows={3}
      />
    </div>
  </div>
</label>
`;

exports[`rendering renders correctly with all props 1`] = `
<label
  className="root
        isRootFullWidth
        isRootInFormLayout
        rootLayoutHorizontal
        isRootRequired
        rootSizeLarge
        isRootStateInvalid
        rootVariantFilled"
  htmlFor="test"
  id="test__label"
>
  <div
    className="label
          isLabelHidden"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="field"
  >
    <div
      className="inputContainer"
    >
      <textarea
        className="input"
        cols={10}
        disabled={true}
        id="test"
        onChange={null}
        placeholder="placeholder"
        required={true}
        rows={5}
        value="value"
      />
      <div
        className="bottomLine"
      />
    </div>
    <div
      className="helperText"
      id="test__helperText"
    >
      some help
    </div>
  </div>
</label>
`;

exports[`rendering renders correctly with hidden label 1`] = `
<label
  className="root
        
        
        rootLayoutVertical
        
        rootSizeMedium
        
        rootVariantOutline"
  htmlFor="test"
  id="test__label"
>
  <div
    className="label
          isLabelHidden"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="field"
  >
    <div
      className="inputContainer"
    >
      <textarea
        className="input"
        cols={null}
        disabled={false}
        id="test"
        onChange={null}
        placeholder={null}
        required={false}
        rows={3}
      />
    </div>
  </div>
</label>
`;
